tests/trestle/core/commands/validate_test.py

```diff
@@ -296,7 +303,107 @@ def test_validate_catalog_params(sample_catalog_rich_controls: Catalog) -> None:
+
+    new_ssp.import_profile.href = original_href
+    ModelUtils.save_top_level_model(new_ssp, tmp_trestle_dir, ssp_name, FileContentType.JSON)
+
+
+def test_rule_param_values_validator_unhappy(tmp_trestle_dir: pathlib.Path, monkeypatch: MonkeyPatch) -> None:
+    """Test validation of rules in SSP."""
+    args = argparse.Namespace(mode=const.VAL_MODE_RULES)
+    validator: Validator = validator_factory.get(args)
+    prof_name = 'comp_prof'
+    ssp_name = 'my_ssp'
+
+    gen_args, _ = setup_for_ssp(tmp_trestle_dir, prof_name, ssp_name)
+    args_compdefs = gen_args.compdefs
+    test_utils.gen_and_assemble_first_ssp(prof_name, ssp_name, gen_args, tmp_trestle_dir, monkeypatch)
+
+    orig_ssp, _ = ModelUtils.load_model_for_class(tmp_trestle_dir, ssp_name, ossp.SystemSecurityPlan)
+    # starts editing by grabbing by components from second imp req
+    imp_reqs = orig_ssp.control_implementation.implemented_requirements
+    imp_req = next((i_req for i_req in imp_reqs if i_req.control_id == 'ac-3'), None)
+    by_components = imp_req.by_components
+    # changes values for set parameter to test inequality
+    by_components[0].set_parameters[1].values = ['shared_param_1_ab_opt_3']
+
+    ModelUtils.save_top_level_model(orig_ssp, tmp_trestle_dir, ssp_name, FileContentType.JSON)
+    assert not validator.model_is_valid(orig_ssp, True, tmp_trestle_dir)
+
+    # test a by_component statement param value is added and
+    # reassemble ssp
+    ssp_assemble = f'trestle author ssp-assemble -m {ssp_name} -o {ssp_name} -cd {args_compdefs}'
+    test_utils.execute_command_and_assert(ssp_assemble, 0, monkeypatch)
+    # load new ssp
+    new_ssp, _ = ModelUtils.load_model_for_class(tmp_trestle_dir, ssp_name, ossp.SystemSecurityPlan)
+    # grabs by component elements at statement level
+    imp_reqs = new_ssp.control_implementation.implemented_requirements
+    imp_req_ac = next((i_req for i_req in imp_reqs if i_req.control_id == 'ac-1'), None)
+    by_components = imp_req_ac.statements[0].by_components
+    by_components[1].set_parameters = []
+
+    new_set_parameter = gens.generate_sample_model(ossp.SetParameter)
+    new_set_parameter.values = ['shared_param_1_ab_opt_3']
+    new_set_parameter.param_id = 'shared_param_1'
+    # addes new value to current set parameter values
+    # adds a new set parameter to set parameters array for current by component in statement
+    by_components[1].set_parameters.append(new_set_parameter)
+    imp_reqs = new_ssp.control_implementation.implemented_requirements
+    imp_req_at = next((i_req for i_req in imp_reqs if i_req.control_id == 'at-1'), None)
+    by_components = imp_req_at.statements[0].by_components
+    by_components[0].set_parameters = []
+    new_set_parameter_one = gens.generate_sample_model(ossp.SetParameter)
+    new_set_parameter_one.values = ['shared_param_1_ab_opt_1']
+    new_set_parameter_one.param_id = 'shared_param_1'
+    by_components[0].set_parameters.append(new_set_parameter_one)
```

Comment:
There are multiple reuses of variable names here: by_components, imp_reqs, etc.  I think they are ok but I would keep the new_set_parameters separate.  Pls call the first one new_set_parameter_a and the other new_set_parameter_b

Reply:
done